Add code review guidelines (#17)
bitwarden/contributing-docs: docs/contributing/pull-requests/code-review.md

# Code Review Guidelines

At Bitwarden, we encourage everyone to participate in code reviews. A pod will focus primarily on
their own code reviews, but if you see something interesting, feel free to jump in and discuss.

To have efficient code reviews there are a few things to keep in mind (from
[Best Practices for Code Review | SmartBear](https://smartbear.com/learn/code-review/best-practices-for-peer-code-review/)):

- As an author, keep your PRs small - ideally less than 400 lines.
  - This can be tricky in our code base since many things are tightly coupled.
  - Consider splitting up a PR into multiple smaller PRs to encourage easier and better quality
    reviews.
  - Target `master` if the code is functionally complete, otherwise target a feature branch.
- Take your time when reviewing - expect a rate of less than 500 lines of code per hour.
- Take breaks - don’t review for longer than 60 minutes.

Don’t feel bad for taking your time when doing code reviews! They often take longer than you think,

> and we should ensure to be spend as much time as needed.

Reviewer: ```suggestion
and we should be spending as much time as needed.
```

#### Request changes

Request changes should be used when you believe something **needs** to change prior to the PR
getting merged, as it will prevent someone else from approving the PR before your concerns have been
tackled.

We shouldn’t hesitate to use this status, however we should give clear feedback on what needs to
change for the PR to get approved. Likewise a PR author should not be discouraged by a _request for
changes_, it's simply an indication that changes should be made prior to the PR being merged.

:::warning Discarding reviews

While it’s possible to discard reviews, this should be used sparingly. Whenever possible please
reach out to the reviewer first to ensure their concerns have been resolved. Below are a couple of
scenarios where discarding the reviews are generally seen as accepted.

- The reviewer is out of office for a longer duration and their original feedback has been resolved.
  It’s the responsibility of the new reviewer to ensure the original feedback has been addressed.
- The PR is a hotfix that needs to be deployed quickly, and the reviewer is in a different timezone.

>   The feedback can be addressed in a follow up PR, should it not already have been resolved.

Reviewer: ```suggestion
  The feedback can be addressed in a follow up PR, if it hasn't already been resolved.
```

:::

#### Approve

> Approving a PR means that you have confidence in that the code works, and does what the PR claims.

Reviewer 2: ```suggestion
Approving a PR means that you have confidence in that the code works, and that does what the PR claims.
```

> Either by testing it or from previous domain knowledge.

Reviewer 2: ```suggestion
This can be based on testing the change or on previous domain knowledge.
```

- You have read and understood the all code changes in the PR.

> - Verified the linked Jira issue description matches the changes made in the PR.

Reviewer: ```suggestion
- You have verified that the linked Jira issue description matches the changes made in the PR.
```
